### PR TITLE
HAL added for add(array, scalar)

### PR DESCRIPTION
**Pull request by savuor**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/core/src/arithm.cpp [hunk 1]
```diff
@@ -585,14 +585,19 @@ static bool ocl_arithm_op(InputArray _src1, InputArray _src2, OutputArray _dst,
 
 #endif
 
+typedef int (*ScalarFunc)(const uchar* src, size_t step_src,
+                          uchar* dst, size_t step_dst, int width, int height,
+                          void* scalar, bool scalarIsFirst);
```
Comment (savuor) [inline modules/core/src/arithm.cpp:590]: `usrdata` is not supported, this should be fixed for more common cases

File: modules/core/src/hal_replacement.hpp [hunk 2]
```diff
@@ -98,6 +98,19 @@ inline int hal_ni_sub64f(const double *src1_data, size_t src1_step, const double
...
+
+@param src_data source image data
+@param src_step source image step
+@param dst_data destination image data
+@param dst_step destination image step
+@param width width of the images
+@param height height of the images
+@param scalar_data pointer to scalar value
+*/
+inline int hal_ni_addScalar32f32f(const float *src_data, size_t src_step, float *dst_data, size_t dst_step, int width, int height, const float* scalar_data) { return CV_HAL_ERROR_NOT_IMPLEMENTED; }
```
Comment (vpisarev) [inline modules/core/src/hal_replacement.hpp:113]: if we want to support C1, C3 etc. scalars, we need to add `int nchannels` (check that the name is consistent with other HAL entries)

Comment (savuor) [reply modules/core/src/hal_replacement.hpp:113]: Probably we will add support for more channels in later PRs
For now we leave it just for 1 channel

Comment (vpisarev) [reply modules/core/src/hal_replacement.hpp:113]: ok

**Review by vpisarev — APPROVED** [review]

**Pull request merged**